### PR TITLE
feat: Allow to consume the initialized solution

### PR DESCRIPTION
**Pull request by zepfred**
By definition, the initialized solution is always generated by the phase immediately prior to the first Local Search phase.

This pull request adds a new consumer, allowing the solver to inject the initialized solution when a valid initialization phase is identified.

### CODE REVIEW

**Comment by zepfred** [comment]
The job `Downstream - Timefold Solver Enterprise for Python` is failing because it is not fetching the correct branch version. The fix is included in this pull request, but it needs to be merged to take effect.

File: core/src/main/java/ai/timefold/solver/core/api/solver/SolverJobBuilder.java
Comment (triceo) [inline core/src/main/java/ai/timefold/solver/core/api/solver/SolverJobBuilder.java:76]: This is not what was agreed, is it?
First local search, yes. First custom phase, no. Custom phase may still be initializing.

Also, arguably, the definition deserves a better explanation:

"Sets the consumer of the first initialized solution. First initialized solution is the solution at the end of the last phase that immediately precedes the first local search phase. This solution marks the beginning of actual optimization."

Also, I'd consider changing the name of the concept from "initialized solution" to "first initialized solution", to clearly explain what is the significance of it.

File: core/src/main/java/ai/timefold/solver/core/impl/solver/DefaultSolverJob.java
Comment (triceo) [inline core/src/main/java/ai/timefold/solver/core/impl/solver/DefaultSolverJob.java:298]: This seems to proliferate both to the phase, and to the phase scope. Is there any chance how we can only keep it on one of the two? It feels to me like this change touches more places than it needs to; if the information is available on the phase, or if it is available on the scope, why would it also need to be available on the other?

File: core/src/main/java/ai/timefold/solver/core/impl/solver/ConsumerSupport.java
Comment (triceo) [inline core/src/main/java/ai/timefold/solver/core/impl/solver/ConsumerSupport.java:50]: Why would we need to?

File: core/src/main/java/ai/timefold/solver/core/impl/solver/ConsumerSupport.java
Comment (triceo) [inline core/src/main/java/ai/timefold/solver/core/impl/solver/ConsumerSupport.java:65]: Isn't this a bit over-complicated? Why all the locking? We're calling a simple listener.

I'm not versed in this particular aspect of the code; perhaps @rsynek could explain?

File: core/src/main/java/ai/timefold/solver/core/impl/phase/AbstractPhase.java
Comment (triceo) [inline core/src/main/java/ai/timefold/solver/core/impl/phase/AbstractPhase.java:83]: Arguably this needs a better name. What does it mean to be an "initialization phase"?

I'd be more literal here. I'd call this "triggersFirstInitializedSolutionEvent()". Javadoc can then reference the definition of what that event is.

**Review by triceo — CHANGES_REQUESTED** [review]
Good start! Please see comments inline.

File: core/src/main/java/ai/timefold/solver/core/impl/solver/ConsumerSupport.java
Comment (zepfred) [reply core/src/main/java/ai/timefold/solver/core/impl/solver/ConsumerSupport.java:50]: I don't think we need to sync it, unless I'm missing something.

File: core/src/main/java/ai/timefold/solver/core/api/solver/SolverJobBuilder.java
Comment (zepfred) [reply core/src/main/java/ai/timefold/solver/core/api/solver/SolverJobBuilder.java:76]: I thought we had agreed on the first custom phase as well, which could be a custom local search.

image

File: core/src/main/java/ai/timefold/solver/core/impl/solver/DefaultSolverJob.java
Comment (zepfred) [reply core/src/main/java/ai/timefold/solver/core/impl/solver/DefaultSolverJob.java:298]: Okay. Let me check it.

File: core/src/main/java/ai/timefold/solver/core/api/solver/SolverJobBuilder.java
Comment (triceo) [reply core/src/main/java/ai/timefold/solver/core/api/solver/SolverJobBuilder.java:76]: I believe that the part that you're sharing here was only a by-product of the conversation.
At the bottom of the document, you see proposal C saying that we only use the first Local Search.

We can not use a custom phase, because custom phase _could_ be some sort of local search, but it could also be a custom construction heuristic. The only place where we can be _sure_ is before the first actual local search.

File: core/src/main/java/ai/timefold/solver/core/impl/solver/ConsumerSupport.java
Comment (zepfred) [reply core/src/main/java/ai/timefold/solver/core/impl/solver/ConsumerSupport.java:65]: I agree regarding the complexity. The initial goal was to bring this up for discussion because I don't think there is a need for a synchronous call.

I have simplified the logic and will now wait for comments from @rsynek.

File: core/src/main/java/ai/timefold/solver/core/api/solver/SolverJobBuilder.java
Comment (zepfred) [reply core/src/main/java/ai/timefold/solver/core/api/solver/SolverJobBuilder.java:76]: Okay. It makes sense.

File: core/src/main/java/ai/timefold/solver/core/impl/solver/ConsumerSupport.java
Comment (rsynek) [reply core/src/main/java/ai/timefold/solver/core/impl/solver/ConsumerSupport.java]: Generally speaking, if the consumption happens on the solver thread and it's a one-time event, I don't see any need of locking.

On the other hand, all the BSEs are being consumed on a consumer thread that is paired with the solver thread.

The reason is that consuming the solution is typically an IO operation that can delay the Solver. That being said, I leave it up to you to decide whether we keep such behavior (for a single solution, it might not be such a big deal, after all).

If you decide to move the consumption on the consumer thread, then locking might be necessary. Take a look at the final BSE consumption; the principle should be very similar.

Comment (rsynek) [reply core/src/main/java/ai/timefold/solver/core/impl/solver/ConsumerSupport.java]: The Init event should in theory be always the first one, so the consumer will always grab it. However, in case of real-time planning, there might be a pending BSE waiting for consumption, that's why acquiring the `activeConsumption` lock might be necessary.

After the previous BSE is consumed, the Init event should have the right of way.

Comment (zepfred) [reply core/src/main/java/ai/timefold/solver/core/impl/solver/ConsumerSupport.java]: Very useful information indeed. That makes sense, and I'll address that.

**Comment by triceo** [comment]
@rsynek Can you please review the async-related stuff?

File: core/src/main/java/ai/timefold/solver/core/impl/solver/ConsumerSupport.java
Comment (zepfred) [inline core/src/main/java/ai/timefold/solver/core/impl/solver/ConsumerSupport.java:72]: @rsynek, some tests indicate that the first initialized consumer may not run because the final solution starts and shuts down the executor before its execution.

I believe that may not occur in production because the LS wouldn't finish before the first initialized solution is consumed. Let me know if you think I should update the tests instead.

https://github.com/TimefoldAI/timefold-solver/actions/runs/9668030108/job/26671216009?pr=915
https://github.com/TimefoldAI/timefold-solver/actions/runs/9668030300/job/26671219202?pr=915

File: core/src/main/java/ai/timefold/solver/core/impl/solver/ConsumerSupport.java
Comment (rsynek) [inline core/src/main/java/ai/timefold/solver/core/impl/solver/ConsumerSupport.java:72]: Can the following (in theory) happen?

1. here, we acquire the `firstSolutionConsumption` lock
2. now if the init event comes, it waits on https://github.com/TimefoldAI/timefold-solver/pull/915/files#diff-07ca1817df09fb6ca99e48c7070b6e09d9992a9d31e7b02bd821bf62ecf52dedR55
3. the consumer finishes the final best solution and disposes the consumer thread
4. ` scheduleFirstInitializedSolutionConsumption` then fails due to the executor not accepting further tasks.

The important question is whether the final BSE can come before the init event and intuitively, I would say no.

**Review by rsynek — APPROVED** [review]
Nice job, @zepfred ! 

I described one hypothetical problematic situation, but the question is whether it can even happen.

File: core/src/test/java/ai/timefold/solver/core/api/solver/SolverManagerTest.java
Comment (triceo) [inline core/src/test/java/ai/timefold/solver/core/api/solver/SolverManagerTest.java:269]: This test is actually several independent tests in one; only CH, CH-LS, etc. Let's split them.

**Review by triceo — APPROVED** [review]
LGTM after comments resolved.

File: core/src/main/java/ai/timefold/solver/core/impl/solver/ConsumerSupport.java
Comment (zepfred) [reply core/src/main/java/ai/timefold/solver/core/impl/solver/ConsumerSupport.java:72]: Yes, I agree with you and couldn't imagine a scenario where BSE could occur before the init event. The init event will always be triggered at the end of the custom or CH phase immediately before an LS phase. So, the `consumeFirstInitializedSolution` will acquire the lock before the LS executes, as only one phase will try to lock  `firstSolutionConsumption`.

Comment (triceo) [reply core/src/main/java/ai/timefold/solver/core/impl/solver/ConsumerSupport.java:72]: Devil's advocate here: the final CH (the same phase which triggers the FISE) also triggers a BSE. (FIS == BS most likely) Does this change anything?

Comment (zepfred) [reply core/src/main/java/ai/timefold/solver/core/impl/solver/ConsumerSupport.java:72]: This situation is impossible. A final CH won't try to consume the FISE because it must immediately precede a LS, and its [flag](https://github.com/TimefoldAI/timefold-solver/pull/915/files#diff-a5c58876daf3e8f4f7fc44827a6c525a6a55e36d9491bbeb7a566b70c81317b4R305) wouldn't be enabled.

**Pull request merged**
